modules/programs/firefox.nix

```diff
@@ -40,15 +40,50 @@ let
 
   profilesIni = generators.toINI { } profiles;
 
-  mkUserJs = prefs: extraPrefs: ''
-    // Generated by Home Manager.
-
-    ${concatStrings (mapAttrsToList (name: value: ''
-      user_pref("${name}", ${builtins.toJSON value});
-    '') prefs)}
+  mkUserJs = prefs: extraPrefs: bookmarks:
+    let
+      prefs' = lib.optionalAttrs ({ } != bookmarks) {
+        "browser.bookmarks.file" = toString (firefoxBookmarksFile bookmarks);
+        "browser.places.importBookmarksHTML" = true;
+      } // prefs;
+    in ''
+      // Generated by Home Manager.
+
+      ${concatStrings (mapAttrsToList (name: value: ''
+        user_pref("${name}", ${builtins.toJSON value});
+      '') prefs')}
+
+      ${extraPrefs}
+    '';
 
-    ${extraPrefs}
-  '';
+  firefoxBookmarksFile = bookmarks:
+    let
+      escapeXML = replaceStrings [ ''"'' "'" "<" ">" "&" ] [
+        "&quot;"
+        "&apos;"
+        "&lt;"
+        "&gt;"
+        "&amp;"
+      ];
+      mapper = _: entry: ''
+        <DT><A HREF="${escapeXML entry.url}" ADD_DATE="0" LAST_MODIFIED="0"${
```

Comment:
Hmm, with XML escaping the test now errors with
```
Expected /nix/store/0fa81qjwv1z5mq34jlkb3kgp86jzjhqv-firefox-bookmarks.html to be same as /nix/store/6v6l0d4ixqnywrbzidi52by90i2kix9r-profile-settings-expected-bookmarks.html but were different:
--- actual
+++ expected
@@ -7,6 +7,6 @@
 <H1>Bookmarks Menu</H1>
 <DL><p>
 <DT><A HREF="https://www.kernel.org" ADD_DATE="0" LAST_MODIFIED="0">kernel.org</A>
-<DT><A HREF="https://en.wikipedia.org/wiki/Special:Search?search=%s&amp;go=Go" ADD_DATE="0" LAST_MODIFIED="0" SHORTCUTURL="wiki">wikipedia</A>
+<DT><A HREF="https://en.wikipedia.org/wiki/Special:Search?search=%s&go=Go" ADD_DATE="0" LAST_MODIFIED="0" SHORTCUTURL="wiki">wikipedia</A>

 </p></DL>
```
does it still work OK with the escaped href value or should this be unescaped? 😕

Reply:
To run the test:
```console
$ nix-build -I nixpkgs=channel:nixpkgs-unstable --pure tests -A build.firefox-profile-settings
```

Unfortunately this test is not run by the CI system since it causes too heavy downloads.

Reply:
yes, I did test this with nixopts, I was slightly expecting it to fail, but it does work, I'll fix the test.

Reply:
fixed, thanks and sorry I forgot to test